source/Coop/Mod/Persistence/Party/MovementData.cs

```diff
@@ -139,15 +166,15 @@ public static MovementData ReadMovementData(this RailBitBuffer buffer)
             return new MovementData
             {
                 DefaultBehaviour = (AiBehavior) buffer.ReadByte(),
-                TargetSettlement = (id = buffer.ReadMBGUID()) != MovementState.InvalidIndex ?
-                    (Settlement) MBObjectManager.Instance.GetObject(id) :
-                    null,
-                TargetParty = (id = buffer.ReadMBGUID()) != MovementState.InvalidIndex ?
-                    (MobileParty) MBObjectManager.Instance.GetObject(id) :
-                    null,
+                TargetSettlement = (id = buffer.ReadMBGUID()) != Coop.InvalidId
+                    ? (Settlement) MBObjectManager.Instance.GetObject(id)
```

Comment:
I think it would be better creating our own ObjectManager class maybe `CoopObjectManager `so that we can call it like this

`TargetSettlement = (Settlement) CoopObjectManager.GetObject( buffer.ReadMBGUID() )`

It would be much clearer and we will be avoiding validating each time for the `Coop.InvalidId`, it's just an idea what do you think?

Reply:
Would that `CoopObjectManager` do anything else than checking against `InvalidId`?

I was thinking about whether it would be a good idea to create a shared game object manager that autonomously synchronizes game objects across all clients. Because that is currently unsolved and i'm not sure manual synchronization with patches is the way to go, especially when considering mod support in the future.

Reply:
On that note, have a look at `Coop.Mod.ObjectManagerAdapter` in this PR. It's an adapter to track the lifetime of `MBObject` instances. Currently used to attach our own synchronization state objects onto the game objects. It could be extended into a bit more than just an adapter to include the automatic synchronization of all instances. What do you think?

Reply:
There is a lot of advanced topics in what you've done, at least in what I've seen, and I'm still trying to figure it all out, mainly because I'm working in the `ItemRoster` sync. The idea of `CoopObjectManager` would be to act as a wrapper for the `MBObjectManager`, I'm thinking that it could be used like

`TargetSettlement = CoopObjectManager.GetSettlementObject( buffer.ReadMBGUID() )`

So that if for whatever reason something changes in the `MBObjectManager` the only place to edit would be the wrapper. And the casting wouldn't be necessary.

`Coop.Mod.ObjectManagerAdapter` could be that class if the Adapter refers to the pattern, it should be the same concept. 

About to the automatic sync, for me sounds great, but I'm still struggling to understand the currently synchronization that you've implemented, so there is not much more I can contribute, at least for now.